crates/bevy_pbr/src/material.rs

```diff
@@ -444,3 +431,159 @@ pub fn queue_material_meshes<M: SpecializedMaterial>(
         }
     }
 }
+
+/// Common [`Material`] properties, calculated for a specific material instance.
+pub struct MaterialProperties {
+    /// The [`AlphaMode`] of this material.
+    pub alpha_mode: AlphaMode,
+    /// Add a bias to the view depth of the mesh which can be used to force a specific render order
+    /// for meshes with equal depth, to avoid z-fighting.
+    pub depth_bias: f32,
+}
+
+/// Data prepared for a [`Material`] instance.
+pub struct PreparedMaterial<T: Material> {
+    pub bindings: Vec<OwnedBindingResource>,
+    pub bind_group: BindGroup,
+    pub key: T::Data,
+    pub properties: MaterialProperties,
+}
+
+struct ExtractedMaterials<M: Material> {
+    extracted: Vec<(Handle<M>, M)>,
+    removed: Vec<Handle<M>>,
+}
+
+impl<M: Material> Default for ExtractedMaterials<M> {
+    fn default() -> Self {
+        Self {
+            extracted: Default::default(),
+            removed: Default::default(),
+        }
+    }
+}
+
+/// Stores all prepared representations of [`Material`] assets for as long as they exist.
+pub type RenderMaterials<T> = HashMap<Handle<T>, PreparedMaterial<T>>;
+
+/// This system extracts all created or modified assets of the corresponding [`Material`] type
+/// into the "render world".
+fn extract_materials<M: Material>(
+    mut commands: Commands,
+    mut events: EventReader<AssetEvent<M>>,
+    assets: Res<Assets<M>>,
+) {
+    let mut changed_assets = HashSet::default();
+    let mut removed = Vec::new();
+    for event in events.iter() {
+        match event {
+            AssetEvent::Created { handle } | AssetEvent::Modified { handle } => {
+                changed_assets.insert(handle);
+            }
+            AssetEvent::Removed { handle } => {
+                changed_assets.remove(handle);
+                removed.push(handle.clone_weak());
+            }
+        }
+    }
+
+    let mut extracted_assets = Vec::new();
+    for handle in changed_assets.drain() {
+        if let Some(asset) = assets.get(handle) {
+            extracted_assets.push((handle.clone_weak(), asset.clone()));
+        }
+    }
+
+    commands.insert_resource(ExtractedMaterials {
+        extracted: extracted_assets,
+        removed,
+    });
```

Comment:
Yes, but given that loading new materials will be a relatively uncommon / burstey operation (ex: load the materials at the start of the level), its unclear if leaving that vec around is "worth it", as it could claim a lot of space indefinitely.

Its also worth calling out that this is how we handle RenderAssets (this code was adapted from the render asset impl). And iirc we rolled with that impl for similar reasons?

Reply:
Yup. Fair. I’m fine with this as-is.